HLTrigger/JetMET/plugins/HLTHFJetCleaner.cc

```diff
@@ -0,0 +1,268 @@
+  float jetEtaMin_;
+  float jetEtaMax_;
+
+  float sigmaEtaPhiDiffThresh_;
+  float centralEtaStripSizeThresh_;
+  bool applyCornerCut_;
+  bool applySigmaEtaPhiCut_;
+  bool applyStripSizeCut_;
+
+  //virtual void beginRun(edm::Run const&, edm::EventSetup const&) override;
+  //virtual void endRun(edm::Run const&, edm::EventSetup const&) override;
+  //virtual void beginLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&) override;
+  //virtual void endLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&) override;
+
+  // ----------member data ---------------------------
+};
+
+//
+// constants, enums and typedefs
+//
+
+//
+// static data member definitions
+//
+
+//
+// constructors and destructor
+//
+HLTHFJetCleaner::HLTHFJetCleaner(const edm::ParameterSet& iConfig)
+    : jetToken_(consumes<edm::View<reco::PFJet>>(iConfig.getParameter<edm::InputTag>("jets"))),
+      metToken_(consumes<edm::View<reco::CaloMET>>(iConfig.getParameter<edm::InputTag>("mets"))),
+      sigmaEtaEtaToken_(consumes<edm::ValueMap<float>>(iConfig.getParameter<edm::InputTag>("sigmaEtaEta"))),
+      sigmaPhiPhiToken_(consumes<edm::ValueMap<float>>(iConfig.getParameter<edm::InputTag>("sigmaPhiPhi"))),
+      centralEtaStripSizeToken_(
+          consumes<edm::ValueMap<int>>(iConfig.getParameter<edm::InputTag>("centralEtaStripSize"))),
+      jetPtThresh_(iConfig.getParameter<double>("jetPtThresh")),
+      dphiJetMetThresh_(iConfig.getParameter<double>("dphiJetMetThresh")),
+      jetEtaMin_(iConfig.getParameter<double>("jetEtaMin")),
+      jetEtaMax_(iConfig.getParameter<double>("jetEtaMax")),
+      sigmaEtaPhiDiffThresh_(iConfig.getParameter<double>("sigmaEtaPhiDiffThresh")),
+      centralEtaStripSizeThresh_(iConfig.getParameter<int>("centralEtaStripSizeThresh")),
+      applyCornerCut_(iConfig.getParameter<bool>("applyCornerCut")),
+      applySigmaEtaPhiCut_(iConfig.getParameter<bool>("applySigmaEtaPhiCut")),
+      applyStripSizeCut_(iConfig.getParameter<bool>("applyStripSizeCut")) {
+  produces<JetCollection>();
+}
+
+HLTHFJetCleaner::~HLTHFJetCleaner() {
+  // do anything here that needs to be done at destruction time
+  // (e.g. close files, deallocate resources etc.)
+  //
+  // please remove this method altogether if it would be left empty
+}
+
+//
+// member functions
+//
+
+template <typename T>
+std::vector<T> HLTHFJetCleaner::extractValueMapParameters(edm::Handle<edm::ValueMap<T>> valueMapHandle,
+                                                          size_t numJets) {
+  // Extract the values as a vector from a given ValueMap object.
+  std::vector<T> values;
+  values.reserve(numJets);
+  for (auto it = valueMapHandle->begin(); it != valueMapHandle->end(); it++) {
+    auto productId = it.id();
+    for (size_t i = 0; i < numJets; i++) {
+      T value = valueMapHandle->get(productId, i);
+      values.push_back(value);
+    }
+  }
+  return values;
+}
```

Comment:
It would be simpler and safer to access the ValueMap via a reference to the object itself, e.g. `vmap[myJetRef]`, also avoiding additional work with indices, and not having to assume that the order of the objects is the same as the order of the VMap entries.

This is a random example: https://github.com/cms-sw/cmssw/blob/6d2f66057131baacc2fcbdd203588c41c885b42c/RecoMET/METProducers/src/BeamHaloSummaryProducer.cc#L158

Reply:
Thanks for the comment, indeed it would be simpler and safer. Can I ask something though, how would you recommend to obtain `myJetRef`? I tried the following as a first approach but it didn't work:

```cpp
for (size_t ijet = 0; ijet < jets->size(); ijet++) {
    reco::PFJet& jet = (*jets)[ijet]; // Raises compiler error
    float sigmaEtaEta = (*sigmaEtaEtas)[jet];
}
```

I also had some problems with `edm::Ref` (following the example you sent). Is it possible if you can help with that? Thanks so much!

Reply:
I haven't tested this, but here is what I would try.

```cpp
// in the class body
using JetCollection = std::vector<jet_type>;
using JetRef = edm::Ref<JetCollection>;

[..]
// in ::produce

auto const h_jets = iEvent.getHandle(jetsToken_); // edm::Handle

for(uint idx=0; idx<h_jets->size(); ++idx){
  JetRef const jetRef(h_jets, idx);
  cout << vmap[jetRef] << endl; // example
}
```

Reply:
Thanks, I will try this approach.

Reply:
Thanks for the help, this approach worked and I implemented this in [3ca0032](https://github.com/cms-sw/cmssw/pull/37539/commits/3ca0032ab77b75be849b9a3a603a52e6d16ebdf0). 

I just had to change the type of jets handle: `edm::Handle<edm::View>` to `edm::Handle<std::vector>` to make the `edm::Ref` instantiation work. I also removed the now-redundant `extractValueMapParameters` function. Let me know if you want any changes, thanks!